File: app/src/main/java/com/stedi/randomimagegenerator/app/view/fragments/ChooseGeneratorFragment.java
package com.stedi.randomimagegenerator.app.view.fragments;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.stedi.randomimagegenerator.app.R;
import com.stedi.randomimagegenerator.app.di.Components;
import com.stedi.randomimagegenerator.app.model.data.GeneratorType;
import com.stedi.randomimagegenerator.app.other.logger.Logger;
import com.stedi.randomimagegenerator.app.presenter.interfaces.ChooseGeneratorPresenter;
import com.stedi.randomimagegenerator.app.view.adapters.GeneratorTypeAdapter;
import com.stedi.randomimagegenerator.app.view.adapters.GeneratorTypeAdapterImageLoader;
import com.stedi.randomimagegenerator.app.view.dialogs.SimpleIntegerParamsDialog;
import com.stedi.randomimagegenerator.app.view.fragments.base.StepFragment;

import javax.inject.Inject;

import butterknife.BindView;

public class ChooseGeneratorFragment extends StepFragment implements
        ChooseGeneratorPresenter.UIImpl,
        GeneratorTypeAdapter.ClickListener {

    @Inject ChooseGeneratorPresenter presenter;
    @Inject Logger logger;
    @Inject GeneratorTypeAdapterImageLoader adapterImageLoader;

    @BindView(R.id.choose_generator_fragment_recycler_view) RecyclerView recyclerView;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Components.getGenerationComponent(this).inject(this);
        presenter.onAttach(this);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        return inflater.inflate(R.layout.choose_generator_fragment, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        recyclerView.setLayoutManager(new GridLayoutManager(getContext(), 2));
        recyclerView.setHasFixedSize(true);
        presenter.getGeneratorTypes();
    }

    @Override
    public void showTypes(@NonNull GeneratorType[] types, @NonNull GeneratorType selectedType) {
        recyclerView.setAdapter(new GeneratorTypeAdapter(adapterImageLoader, types, selectedType, this));
    }

    @Override
    public void onSelected(@NonNull GeneratorType type) {
        presenter.chooseGeneratorType(type);
    }

    @Override
    public void onEditSelected() {
        presenter.editChoseGeneratorParams();
    }

    @Override
    public void showEditGeneratorParams(@NonNull GeneratorType type) {
        switch (type) {
            case COLORED_PIXELS:
            case COLORED_CIRCLES:
            case COLORED_RECTANGLE:
                SimpleIntegerParamsDialog.newInstance(type).show(getFragmentManager(), SimpleIntegerParamsDialog.class.getSimpleName());
                break;
            case COLORED_NOISE:
                throw new IllegalStateException("todo");
            default:
                throw new IllegalStateException("incorrect behavior");
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenter.onDetach();
    }
}

File: app/src/main/java/com/stedi/randomimagegenerator/app/view/fragments/ChooseGeneratorFragment.kt
package com.stedi.randomimagegenerator.app.view.fragments

import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import butterknife.BindView
import com.stedi.randomimagegenerator.app.R
import com.stedi.randomimagegenerator.app.model.data.GeneratorType
import com.stedi.randomimagegenerator.app.presenter.interfaces.ChooseGeneratorPresenter
import com.stedi.randomimagegenerator.app.view.adapters.GeneratorTypeAdapter
import com.stedi.randomimagegenerator.app.view.components.BaseViewModel
import com.stedi.randomimagegenerator.app.view.dialogs.ColoredNoiseParamsDialog
import com.stedi.randomimagegenerator.app.view.dialogs.SimpleIntegerParamsDialog
import com.stedi.randomimagegenerator.app.view.fragments.base.GenerationFragment
import javax.inject.Inject

class ChooseGeneratorFragmentModel : BaseViewModel<ChooseGeneratorFragment>() {
    @Inject lateinit var presenter: ChooseGeneratorPresenter

    override fun onCreate(view: ChooseGeneratorFragment) {
        view.generationComponent.inject(this)
    }

    override fun onCleared() {
        super.onCleared()
        presenter.onDestroy()
    }
}

class ChooseGeneratorFragment : GenerationFragment(), ChooseGeneratorPresenter.UIImpl, GeneratorTypeAdapter.ClickListener {

    private lateinit var viewModel: ChooseGeneratorFragmentModel

    @BindView(R.id.choose_generator_fragment_recycler_view) lateinit var recyclerView: RecyclerView

    private lateinit var adapter: GeneratorTypeAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        viewModel = ViewModelProviders.of(this).get(ChooseGeneratorFragmentModel::class.java)
        viewModel.init(this)

        viewModel.presenter.onAttach(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreateView(inflater, container, savedInstanceState)
        return inflater.inflate(R.layout.choose_generator_fragment, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        recyclerView.layoutManager = GridLayoutManager(context, resources.getInteger(R.integer.generator_type_adapter_grid_items))
        recyclerView.setHasFixedSize(true)
        adapter = GeneratorTypeAdapter(activity!!, this, false)
        recyclerView.adapter = adapter
        viewModel.presenter.getGeneratorTypes()
    }

    override fun showTypes(types: Array<GeneratorType>, selectedType: GeneratorType) {
        adapter.set(types, selectedType, null)
    }

    override fun onSelected(type: GeneratorType) {
        viewModel.presenter.chooseGeneratorType(type)
    }

    override fun onEditSelected() {
        viewModel.presenter.editChoseGeneratorParams()
    }

    override fun showEditGeneratorParams(type: GeneratorType) {
        val fragmentManager = fragmentManager ?: return
        when (type) {
            in SimpleIntegerParamsDialog.SUPPORTED_TYPES -> {
                SimpleIntegerParamsDialog.newInstance(type).show(fragmentManager)
            }
            GeneratorType.COLORED_NOISE -> {
                ColoredNoiseParamsDialog().show(fragmentManager)
            }
            else -> throw IllegalStateException("incorrect behavior")
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        viewModel.presenter.onDetach()
    }

    override fun onDeselected() {}
}